blackjax/adaptation/chees_adaptation.py

```diff
@@ -0,0 +1,426 @@
+            initial_step_size,
+            optim,
+            num_warmup_steps,
+        )
+        kernel = blackjax.dynamic_hmc(logprob_fn, **parameters).step
+        new_states, info = jax.vmap(kernel)(key_sample, last_states)
+
+    Parameters
+    ----------
+    logprob_fn
+        The log density probability density function from which we wish to sample.
+    num_chains
+        Number of chains used for cross-chain warm-up training.
+    jitter_generator
+        Optional function that generates a value in [0, 1] used to jitter the trajectory
+        lengths given a PRNGKey, used to propose the number of integration steps. If None,
+        then a quasi-random Halton is used to jitter the trajectory length.
+
+    Returns
+    -------
+    A function that returns the last cross-chain state, a sampling kernel with the
+    tuned parameter values, and all the warm-up states for diagnostics.
+
+    """
+
+    def run(
+        rng_key: PRNGKey,
+        positions: ArrayLikeTree,
+        step_size: float,
+        optim: optax.GradientTransformation,
+        num_steps: int = 1000,
+        *,
+        max_sampling_steps: int = 1000,
+    ):
+        num_dim = 0
+        for d in jax.tree_util.tree_leaves(
+            jax.tree_util.tree_map(
+                lambda p: p.reshape(num_chains, -1).shape[1], positions
+            )
+        ):
+            num_dim += d
+
+        key_init, key_step = jax.random.split(rng_key)
+
+        if jitter_generator is not None:
+            jitter_gn = jitter_generator
+            next_random_arg_fn = lambda key: jax.random.split(key)[1]
+            init_random_arg = key_init
+        else:
+            jitter_gn = lambda i: _halton_sequence(
+                i, np.ceil(np.log2(num_steps + max_sampling_steps))
+            )
+            next_random_arg_fn = lambda i: i + 1
+            init_random_arg = 0
+
+        def integration_steps_fn(random_generator_arg, trajectory_length_adjusted):
+            return jnp.ceil(
+                jitter_gn(random_generator_arg) * trajectory_length_adjusted
+            )
+
+        step_fn = hmc.build_dynamic_kernel(
+            next_random_arg_fn=next_random_arg_fn,
+            integration_steps_fn=integration_steps_fn,
+        )
+
+        init, update = base(jitter_gn, next_random_arg_fn, optim)
+
+        def one_step(carry, rng_key):
+            states, adaptation_state = carry
+
+            def one_step(rng_key, state):
+                return step_fn(
+                    rng_key,
+                    state,
+                    logprob_fn,
+                    step_size=adaptation_state.step_size,
+                    inverse_mass_matrix=jnp.ones(num_dim),
+                    trajectory_length_adjusted=adaptation_state.trajectory_length
+                    / adaptation_state.step_size,
+                )
+
+            keys = jax.random.split(rng_key, num_chains)
+            new_states, info = jax.vmap(one_step)(keys, states)
```

Comment:
Could you refactor this to:
```python
keys = jax.random.split(rng_key, num_chains)
jax.vmap(step_fn, in_axis=[0, 0, None...])(keys, states, ...)
```

Reply:
`vmap` can't handle well kwargs, [see](https://github.com/google/jax/issues/7465). Best I could do was `partial` them out.